collect_app/src/main/java/org/odk/collect/android/tasks/InstanceSyncTask.java

```diff
@@ -103,14 +107,22 @@ protected String doInBackground(Void... params) {
                     while (instanceCursor.moveToNext()) {
                         String instanceFilename = instanceCursor.getString(
                                 instanceCursor.getColumnIndex(InstanceColumns.INSTANCE_FILE_PATH));
-                        candidateInstances.remove(instanceFilename);
+                        if (candidateInstances.contains(instanceFilename)) {
+                            candidateInstances.remove(instanceFilename);
+                        } else {
+                            filesToRemove.add(instanceFilename);
+                        }
                     }
+
                 } finally {
                     if (instanceCursor != null) {
                         instanceCursor.close();
                     }
                 }
 
+                InstancesDao instancesDao = new InstancesDao();
+                instancesDao.deleteInstancesFromIDs(filesToRemove.toArray(new String[filesToRemove.size()]));
```

Comment:
I think you can change the parameter to list since you're gonna use the array directly (you're going to reassign the values from the array to another array).